### PR TITLE
BatchDiagGGNMC Extension

### PR DESCRIPTION
**Pull request by sbharadwajj**
https://github.com/fKunstner/backpack-discuss/issues/81

### CODE REVIEW

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 1]
```diff
@@ -196,3 +196,28 @@ def __init__(self):
             loss_hessian_strategy=LossHessianStrategy.EXACT,
             savefield="diag_ggn_exact_batch",
         )
+
+
+class BatchDiagGGNMC(BatchDiagGGN):
+    """
+    Diagonal of the Generalized Gauss-Newton/Fisher.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:203]: Suggestion: "Diagonal" -> "Individual diagonal"

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 2]
```diff
@@ -196,3 +196,28 @@ def __init__(self):
...
+    Uses a Monte-Carlo approximation of
+    the Hessian of the loss w.r.t. the model output.
+
+    Stores the output in :code:`diag_ggn_mc_batch` as a ``[N x ...]`` tensor,
+    where ``N`` batch size and ``...`` is the shape of the gradient.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:208]: Missing "is the"

File: test/extensions/implementation/base.py [hunk 3]
```diff
@@ -32,6 +32,10 @@ def diag_ggn_mc(self, mc_samples):
         """MC approximation of Diagonal of Gauss Newton"""
         raise NotImplementedError
 
+    def diag_ggn_mc_batch(self, mc_samples):
+        """MC approximation of Individual diagonal of Generalized Gauss-Newton/Fisher"""
```
Comment (f-dangel) [inline test/extensions/implementation/base.py:36]: "Individual" -> "individual"

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 4]
```diff
@@ -89,3 +90,46 @@ def test_diag_ggn_mc(problem):
...
+
+    Args:
+        problem (ExtensionsTestProblem): Problem for extension test.
+    """
+    problem.set_up()
+
+    backpack_res = BackpackExtensions(problem).diag_ggn_exact_batch()
+    # NOTE May crash for large networks because of large number of samples.
+    # If necessary, resolve by chunking samples into smaller batches + averaging
+    mc_samples = 100000
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:129]: I have experienced this now when executing the tests on GPU, and [fixed](https://github.com/f-dangel/backpack/commit/9ad9ee8e89f0a9445f4acca0f34371ca0e1ff383) it for `DiagGGNMC`. The chunking is implemented [here](https://github.com/f-dangel/backpack/blob/branching/test/extensions/secondorder/diag_ggn/test_diag_ggn.py#L92-L94). 

Please also implement the chunking for `DiagGGNMCBatch`, such that these tests can be executed on a GPU.

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 5]
```diff
@@ -43,6 +43,7 @@ def test_diag_ggn_batch(problem):
 
 
 MC_ATOL = 1e-4
+MC_BATCH_ATOL = 1e-2
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:46]: This value is large. I am concerned that we would not catch failures with it.

Can we at least test that the MC approximation becomes better as we use more samples? I.e. the test below with 1000 samples should require larger tolerances to pass than the test with 100.000 samples?

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 6]
```diff
@@ -89,3 +90,47 @@ def test_diag_ggn_mc(problem):
...
+    mc_samples = 300000
+    chunks = 30
+    backpack_res_mc_avg = BackpackExtensions(problem).diag_ggn_mc_batch_chunk(
+        mc_samples, chunks=chunks
+    )
+
+    check_sizes_and_values(
+        backpack_res, backpack_res_mc_avg, atol=MC_BATCH_ATOL, rtol=MC_RTOL
+    )
+    problem.tear_down()
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:136]: Could you move those tests to `test_batch_diag_ggn`?

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 7]
```diff
@@ -89,3 +90,47 @@ def test_diag_ggn_mc(problem):
...
+    check_sizes_and_values(
+        backpack_res, backpack_res_mc_avg, atol=MC_BATCH_ATOL, rtol=MC_LIGHT_RTOL
+    )
+    problem.tear_down()
+
+
+@pytest.mark.montecarlo
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_mc_batch(problem):
+    """Test the MC approximation of Diagonal of Gauss-Newton
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:118]: add 'individual'

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 8]
```diff
@@ -89,3 +90,47 @@ def test_diag_ggn_mc(problem):
...
+
+    Args:
+        problem (ExtensionsTestProblem): Problem for extension test.
+    """
+    problem.set_up()
+
+    backpack_res = BackpackExtensions(problem).diag_ggn_exact_batch()
+    mc_samples = 3000
+    backpack_res_mc_avg = BackpackExtensions(problem).diag_ggn_mc_batch(mc_samples)
+    # sum the batch, and compare with exact mc
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:108]: Is this really done?

Comment (sbharadwajj) [reply test/extensions/secondorder/diag_ggn/test_diag_ggn.py:108]: Ah, do you mean the `# sum the batch, and compare with exact mc`? I think I've missed it. Will look into it now

File: test/extensions/secondorder/diag_ggn/test_batch_diag_ggn.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    problem.set_up()
+
+    backpack_res = BackpackExtensions(problem).diag_ggn_exact_batch()
+    autograd_res = AutogradExtensions(problem).diag_ggn_batch()
+
+    check_sizes_and_values(autograd_res, backpack_res)
+    problem.tear_down()
+
+
+MC_BATCH_ATOL = 1e-4
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_batch_diag_ggn.py:29]: This can be `MC_ATOL`

**Pull request merged**